src/core/loopback-setup.c

```diff
@@ -67,23 +175,51 @@ static bool check_loopback(sd_netlink *rtnl) {
 
 int loopback_setup(void) {
         _cleanup_(sd_netlink_unrefp) sd_netlink *rtnl = NULL;
+        struct state state = {};
         int r;
 
         r = sd_netlink_open(&rtnl);
         if (r < 0)
-                return r;
+                return log_error_errno(r, "Failed to open netlink: %m");
+
+        /* Note that we add the IP addresses here explicitly even though the kernel does that too implicitly when
+         * setting up the loopback device. The reason we do this here a second time (and possibly race against the
+         * kernel) is that we want to synchronously wait until the IP addresses are set up correctly, see
+         *
+         * https://github.com/systemd/systemd/issues/5641 */
+
+        r = add_ipv4_address(rtnl, &state);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enqeue IPv4 loopback address add request: %m");
+
+        r = add_ipv6_address(rtnl, &state);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enqeue IPv4 loopback address add request: %m");
+
+        r = start_loopback(rtnl, &state);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enqeue loopback interface start request: %m");
+
+        while (state.n_messages > 0) {
+                r = sd_netlink_wait(rtnl, USEC_INFINITY);
```

Comment:
uh, dunno, i'd be quite surprised if the kernel wouldn't answer here... but I figure it's also OK to shorten the timeout, dunno